docs/reference/constrained.rst file added.

peterhinch opened the pull request:
This offers a tutorial on running MicroPython on resource constrained devices such as microcontrollers. It aims to be platform independent.

It may also be reviewed [here](http://hinch.me.uk/html/reference/constrained.html).

The doc contains three questions to the reviewers (search on TODO) where there are gaps in my knowledge. I'll fix these as soon as answers are forthcoming :)

All comments especially on factual errors and omissions welcome.

peterhinch committed:
docs/reference/constrained.rst file added.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,415 @@
+Symbol Meaning
+====== =================
+   .   free block
+   =   ??? *TODO reviewer please advise. Also is this list complete?*
```
Tail 't' has been changed to '='.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,415 @@
+
+The ``gc.collect()`` method returns an integer.
+
+*TODO reviewer please advise what it signifies*
```
It's the number of distinct memory regions that were reclaimed in the collection (more precisely, the number of heads that were turned into frees).  Only unix/windows return an integer, the bare-metal ports do not (for efficiency reasons).

peterhinch committed:
constrained.rst Reviewer comments incorporated

peterhinch commented:
@dpgeorge Apologies, I somehow missed your prompt response to this. I've now updated the doc.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+
+.. code::
+
+    ROWS = const(33)
```
Please add `from micropython import const`; that is the new way to do it to be compatible with CPython.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+In both instances where the constant is assigned to a variable the compiler
+will avoid coding a lookup to the name of the constant by substituting its
+literal value. This saves bytecode and hence RAM. However the ``ROWS`` value
+will occupy a single machine word because another module might import it. This
```
It'll occupy at least 2 machine words, one for the key, one for the value, in the globals dict.

dpgeorge replied:
"might import or use it"

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+literal value. This saves bytecode and hence RAM. However the ``ROWS`` value
+will occupy a single machine word because another module might import it. This
+RAM can be saved by prepending the name with an underscore as in ``_COLS``: this
+symbol is not available for import so will not occupy RAM.
```
"this symbol is not visible outside the module so will not occupy RAM"

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+symbol is not available for import so will not occupy RAM.
+
+The argument to ``const()`` may be anything which, at compile time, evaluates
+to an integer e.g. ``0x100`` or ``1 << 8``.
```
"It can even include other const symbols that have already been defined, eg ``1 << BIT``."

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+all modules have been imported. This maximises the RAM available to the
+compiler.
+
+If RAM is still insufficient to compile all modules one solution is to
```
A general question: do you often hit ram limitations when importing modules? #2417 discusses how this can be helped (but it's not relevant for this document).

peterhinch replied:
I've encountered this with larger modules, I think it occurs on the order of 1K LOC on the Pyboard.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+Python built-in types.
+
+In its current state of development the compiler will not do this for other
+immutable types, which are copied to RAM on importation.
```
str, bytes, int, float and complex are all frozen to flash.  The only one missing is tuples of constant data, and hopefully that'll be implemented one day :)

You could in principle store arbitrary data in a big int.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+
+**String concatenation**
+
+Consider the following code fragments aimed at producing constant strings:
```
"aim to produce"

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+    buf = bytearray(100)
+    while True:
+        i2c.recv(buf, 10)
+        # process data in buf
```
It'd probably be better to give an example using the new machine API.  machine.SPI is now fully implemented on pyboard (and pyb.SPI is mostly forwards compatible) so you can do:
```
spi.read(100)
spi.readinto(buf)
```

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+
+.. code::
+
+    spi.send((0x70, 0x03))
```
Does this even work?

peterhinch replied:
Rewritten with (hopefully) a better example.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+string and an array of bytes. Where an application does not require Unicode
+support it saves RAM to use ``bytes`` and ``bytearray`` objects where possible.
+Note that most string methods (e.g. ``strip()``) apply also to ``bytes``
+instances so the process of eliminating Unicode can be painless.
```
Unicode str doesn't take any more RAM than bytes, as long as all chars are ASCII (ie < 126).

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+    print('Func run free: {} allocated: {}'.format(gc.mem_free(), gc.mem_alloc()))
+    gc.collect()
+    print('Garbage collect free: {} allocated: {}, peak {}'.format(
+        gc.mem_free(), gc.mem_alloc(), micropython.mem_peak()))
```
pyboard doesn't have mem_peak function, and you don't talk about it below so I'd remove it.

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+In general modules should instantiate data objects at runtime using constructors
+or other initialisation functions. The reason is that if this occurs on
+initialisation the compiler may be starved of RAM when subsequent modules are
+imported. If modules do instantiate data on import then ``gc,collect()`` issued
```
"gc.collect()"

dpgeorge commented on docs/reference/constrained.rst:
```diff
@@ -0,0 +1,414 @@
+        # do it
+
+In C the second would be inefficient because the string literal would occupy
+RAM in each comparison. Further the ``strcmp`` comparison operates in a
```
I don't understand what you mean by "occupy RAM in each comparison"?

peterhinch replied:
I've rewritten this.

peterhinch committed:
constrained.rst Amended post review.

peterhinch commented:
@dpgeorge Now updated as per your comments. I've made a couple of other changes. I relegated the description of gc.collect() return values to a footnote (it's somewhat OT) and added some final notes on standard libraries which I thought helpful.

dpgeorge commented:
Thank you.  Merged in 9e1dec181806d7cd34b5dfec8cdce4c41b39ab53.
